File: deprecated/source/models/supabase/journey_progress.py
import json
from datetime import datetime
from typing import ClassVar, Dict, Optional
from source.models.supabase.supabase_model import SupabaseModel
from uuid import UUID
from pydantic import Field, field_validator
from supabase.client import AsyncClient


class JourneyProgressModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "journey_progress"
    id: Optional[UUID] = Field(default=None)
    journey_id: UUID
    journey_version_id: UUID
    assigned_at: Optional[datetime] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)
    started_at: Optional[datetime] = Field(default=None)
    completed_at: Optional[datetime] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v


class JourneyProgressLLMConversationMessagesModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "journey_progress_llm_conversation_messages"
    journey_item_progress_id: UUID
    message_id: UUID
    conversation_id: UUID
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    @classmethod
    async def save_to_supabase(
        cls,
        supabase: AsyncClient,
        instance,
        on_conflict=["journey_item_progress_id", "message_id"],
    ):
        return await super(
            JourneyProgressLLMConversationMessagesModel, cls
        ).save_to_supabase(supabase, instance, on_conflict)

    @classmethod
    async def upsert_to_supabase(
        cls,
        supabase: AsyncClient,
        instances,
        on_conflict=["journey_item_progress_id", "message_id"],
        id_column=None,
    ):
        return await super(
            JourneyProgressLLMConversationMessagesModel, cls
        ).upsert_to_supabase(supabase, instances, on_conflict, id_column)

    @classmethod
    async def fetch_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "journey_item_progress_id": value.journey_item_progress_id,
                "message_id": value.message_id,
            }
        return await super().fetch_from_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def exists_in_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "journey_item_progress_id": value.journey_item_progress_id,
                "message_id": value.message_id,
            }
        return await super().exists_in_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def delete_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "journey_item_progress_id": value.journey_item_progress_id,
                "message_id": value.message_id,
            }
        return await super().delete_from_supabase(
            supabase, value=value, id_column=id_column
        )


class JourneyProgressLLMConversationsModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "journey_progress_llm_conversations"
    progress_id: UUID
    conversation_id: UUID
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    @classmethod
    async def save_to_supabase(
        cls,
        supabase: AsyncClient,
        instance,
        on_conflict=["progress_id", "conversation_id"],
    ):
        return await super(JourneyProgressLLMConversationsModel, cls).save_to_supabase(
            supabase, instance, on_conflict
        )

    @classmethod
    async def upsert_to_supabase(
        cls,
        supabase: AsyncClient,
        instance,
        on_conflict=["progress_id", "conversation_id"],
        id_column=None,
    ):
        return await super(
            JourneyProgressLLMConversationsModel, cls
        ).upsert_to_supabase(supabase, instance, on_conflict, id_column)

    @classmethod
    async def fetch_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "progress_id": value.progress_id,
                "conversation_id": value.conversation_id,
            }
        return await super(
            JourneyProgressLLMConversationsModel, cls
        ).fetch_from_supabase(supabase, value=value, id_column=id_column)

    @classmethod
    async def exists_in_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "progress_id": value.progress_id,
                "conversation_id": value.conversation_id,
            }
        return await super().exists_in_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def delete_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {
                "progress_id": value.progress_id,
                "conversation_id": value.conversation_id,
            }
        return await super().delete_from_supabase(
            supabase, value=value, id_column=id_column
        )


class JourneyItemProgressModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "journey_item_progress"
    id: Optional[UUID] = Field(default=None)
    journey_progress_id: UUID
    journey_item_id: UUID
    journey_item_version_id: UUID
    data: Optional[Dict] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)
    started_at: Optional[datetime] = Field(default=None)
    completed_at: Optional[datetime] = Field(default=None)

    @field_validator("data", mode="before")
    def validate_data(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v

File: deprecated/lib_js/models/supabase/journeyProgress.js
import { SupabaseModel } from "./supabaseModel.js";

// Define JourneyProgressModel
export class JourneyProgressModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        journeyId: { type: "uuid", required: true, dbColumn: "journey_id" },
        journeyVersionId: {
            type: "uuid",
            required: true,
            dbColumn: "journey_version_id"
        },
        assignedAt: { type: "date", required: false, dbColumn: "assigned_at" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        },
        startedAt: { type: "date", required: false, dbColumn: "started_at" },
        completedAt: { type: "date", required: false, dbColumn: "completed_at" }
    };
}

// Define JourneyProgressLLMConversationMessagesModel
export class JourneyProgressLLMConversationMessagesModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress_llm_conversation_messages";
    static TABLE_FIELDS = {
        journeyItemProgressId: {
            type: "uuid",
            required: true,
            dbColumn: "journey_item_progress_id"
        },
        messageId: { type: "uuid", required: true, dbColumn: "message_id" },
        conversationId: {
            type: "uuid",
            required: true,
            dbColumn: "conversation_id"
        },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["journeyItemProgressId", "messageId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instances,
        onConflict = ["journeyItemProgressId", "messageId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(
            supabase,
            instances,
            onConflict,
            idColumn
        );
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId: value.attributes.journeyItemProgressId,
                messageId: value.attributes.messageId
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId: value.attributes.journeyItemProgressId,
                messageId: value.attributes.messageId
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId: value.attributes.journeyItemProgressId,
                messageId: value.attributes.messageId
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

// Define JourneyProgressLLMConversationsModel
export class JourneyProgressLLMConversationsModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress_llm_conversations";
    static TABLE_FIELDS = {
        progressId: { type: "uuid", required: true, dbColumn: "progress_id" },
        conversationId: {
            type: "uuid",
            required: true,
            dbColumn: "conversation_id"
        },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["progressId", "conversationId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instance,
        onConflict = ["progressId", "conversationId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(supabase, instance, onConflict, idColumn);
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId,
                conversationId: value.attributes.conversationId
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId,
                conversationId: value.attributes.conversationId
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId,
                conversationId: value.attributes.conversationId
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

// Define JourneyItemProgressModel
export class JourneyItemProgressModel extends SupabaseModel {
    static TABLE_NAME = "journey_item_progress";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        journeyProgressId: {
            type: "uuid",
            required: true,
            dbColumn: "journey_progress_id"
        },
        journeyItemId: {
            type: "uuid",
            required: true,
            dbColumn: "journey_item_id"
        },
        journeyItemVersionId: {
            type: "uuid",
            required: true,
            dbColumn: "journey_item_version_id"
        },
        data: { type: "json", required: false, dbColumn: "data" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        },
        startedAt: { type: "date", required: false, dbColumn: "started_at" },
        completedAt: { type: "date", required: false, dbColumn: "completed_at" }
    };
}
